File: java-boot/src/test/java/se/callistaenterprise/kotlindsl/repo/OrderRepositoryTest.java
package se.callistaenterprise.kotlindsl.repo;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import se.callistaenterprise.kotlindsl.data.DataRepositories;
import se.callistaenterprise.kotlindsl.dsl.OrderDsl;
import se.callistaenterprise.kotlindsl.model.Item;
import se.callistaenterprise.kotlindsl.model.OrderIn;
import se.callistaenterprise.kotlindsl.testutil.DslIntegrationTest;
import se.callistaenterprise.kotlindsl.testutil.DslJooqTest;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@DslJooqTest
public class OrderRepositoryTest extends DslIntegrationTest {

    private final DataRepositories repos;
    private final OrderRepository orderRepo;
    private final CustomerRepository customerRepo;
    private final ProductRepository productRepo;

    public OrderRepositoryTest(
            @Autowired DataRepositories repos,
            @Autowired OrderRepository orderRepo,
            @Autowired CustomerRepository customerRepo,
            @Autowired ProductRepository productRepo) {
        this.repos = repos;
        this.orderRepo = orderRepo;
        this.customerRepo = customerRepo;
        this.productRepo = productRepo;
    }

    @Test
    public void roundTripWorks() {
        var dsl = new OrderDsl();
        var customerDsl = dsl.customer("12345");
        var productDsl1 = dsl.product("A product");
        var productDsl2 = dsl.product("Another product");
        repos.persist(dsl);

        var customer = customerRepo.getById(customerDsl.getId());
        var product1 = productRepo.getById(productDsl1.getId());
        var product2 = productRepo.getById(productDsl2.getId());

        var orderIn = new OrderIn("1234", customer, Arrays.asList(new Item(product1, 2L), new Item(product2, 3L)));
        var id = orderRepo.create(orderIn);
        var order = orderRepo.getById(id);

        assertEquals(order.orderId(), orderIn.orderId());
        assertEquals(order.customer(), orderIn.customer());
        assertTrue(order.items().containsAll(orderIn.items()));
        assertTrue(orderIn.items().containsAll(order.items()));
    }
}

File: kotlin-boot/src/test/kotlin/se/callistaenterprise/kotlindsl/repo/OrderRepositoryTest.kt
package se.callistaenterprise.kotlindsl.repo

import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import se.callistaenterprise.kotlindsl.data.DataRepositories
import se.callistaenterprise.kotlindsl.dsl.OrderDsl
import se.callistaenterprise.kotlindsl.model.Item
import se.callistaenterprise.kotlindsl.model.OrderIn
import se.callistaenterprise.kotlindsl.testutil.DslIntegrationTest
import se.callistaenterprise.kotlindsl.testutil.DslJooqTest

@DslJooqTest
class OrderRepositoryTest(
    @Autowired val repos: DataRepositories,
    @Autowired val orderRepo: OrderRepository,
    @Autowired val customerRepo: CustomerRepository,
    @Autowired val productRepo: ProductRepository
) : DslIntegrationTest() {

    @Test
    fun `Roundtrip works`() {
        val dsl = OrderDsl()
        val customerDsl = dsl.customer("12345")
        val productDsl1 = dsl.product("A product")
        val productDsl2 = dsl.product("Another product")
        repos.persist(dsl)

        val customer = customerRepo.getById(customerDsl.id)!!
        val product1 = productRepo.getById(productDsl1.id)!!
        val product2 = productRepo.getById(productDsl2.id)!!

        val orderIn = OrderIn("1234", customer, listOf(Item(product1, 2), Item(product2, 3)))
        val id = orderRepo.create(orderIn)
        val order = orderRepo.getById(id)!!

        order.orderId shouldBe orderIn.orderId
        order.customer shouldBe orderIn.customer
        order.items shouldContainExactlyInAnyOrder orderIn.items
    }
}
